Build bug: `Pattern already registered with interpreter`

If these are the case ...
- there is a GIPSL spec with more than one access to a pattern, ...
- one configured GIPS to build the problem in parallel, ...
- one is not lucky, ...

then the following exception may occur when executing the API:
```
Exception in thread "main" java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Pattern already registered with interpreter: org.emoflon.ibex.patternmodel.IBeXPatternModel.impl.IBeXContextPatternImpl@13e1080b (name: occupantRoomDay) (hasCountExpression: false) (documentation: )
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:540)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:567)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:670)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:160)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:765)
	at org.emoflon.gips.core.GipsEngine.lambda$2(GipsEngine.java:115)
	at org.emoflon.gips.core.util.Observer.observe(Observer.java:50)
	at org.emoflon.gips.core.util.Observer.observe(Observer.java:35)
	at org.emoflon.gips.core.GipsEngine.lambda$0(GipsEngine.java:89)
	at org.emoflon.gips.core.util.Observer.observe(Observer.java:50)
	at org.emoflon.gips.core.util.Observer.observe(Observer.java:35)
	at org.emoflon.gips.core.GipsEngine.buildProblemTimed(GipsEngine.java:85)
	at org.emoflon.gips.ihtc.runner.AbstractIhtcGipsRunner.buildAndSolve(AbstractIhtcGipsRunner.java:194)
	at org.emoflon.gips.ihtc.runner.IhtcHardOnlyGipsRunner.run(IhtcHardOnlyGipsRunner.java:73)
	at org.emoflon.gips.ihtc.runner.cli.IhtcGipsHeadlessRunner.execute(IhtcGipsHeadlessRunner.java:103)
	at org.emoflon.gips.ihtc.runner.cli.IhtcGipsHeadlessRunner.main(IhtcGipsHeadlessRunner.java:51)
Caused by: java.lang.IllegalArgumentException: Pattern already registered with interpreter: org.emoflon.ibex.patternmodel.IBeXPatternModel.impl.IBeXContextPatternImpl@13e1080b (name: occupantRoomDay) (hasCountExpression: false) (documentation: )
	at org.emoflon.ibex.gt.engine.GraphTransformationInterpreter.registerGraphTransformationPattern(GraphTransformationInterpreter.java:665)
	at org.emoflon.ibex.gt.api.GraphTransformationPattern.<init>(GraphTransformationPattern.java:73)
	at ihtcgipssolution.hardonly.api.rules.OccupantRoomDayPattern.<init>(OccupantRoomDayPattern.java:37)
	at ihtcgipssolution.hardonly.api.HardonlyAPI.occupantRoomDay(HardonlyAPI.java:287)
	at ihtcgipssolution.hardonly.api.gips.constraint.MappingConstraint5OnroomDayLoad.buildConstantRhs(MappingConstraint5OnroomDayLoad.java:34)
	at ihtcgipssolution.hardonly.api.gips.constraint.MappingConstraint5OnroomDayLoad.buildConstantRhs(MappingConstraint5OnroomDayLoad.java:1)
	at org.emoflon.gips.core.GipsMappingConstraint.buildConstraint(GipsMappingConstraint.java:56)
	at org.emoflon.gips.core.GipsMappingConstraint.lambda$0(GipsMappingConstraint.java:31)
	at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1787)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at org.emoflon.gips.core.GipsMappingConstraint.buildConstraints(GipsMappingConstraint.java:30)
	at org.emoflon.gips.core.GipsEngine.lambda$8(GipsEngine.java:115)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1787)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

Respective GIPSL snippet:
```
// This constraint sets the respective `roomDayGender` mapping to 1
// if and only if the gender `g` is present in room `r` on day `d`.
constraint with roomDayGender {
	mappings.roomDayPatientLoad->filter( //
		element.nodes.d == context.nodes.d //
		& element.nodes.r == context.nodes.r //
		& element.nodes.p.gender == context.nodes.g.name //
	)->sum(element.value)
	+
	patterns.occupantRoomDay->filter( //
		element.nodes.d == context.nodes.d //
		& element.nodes.r == context.nodes.r //
		& element.nodes.o.gender == context.nodes.g.name //
	)->sum(1)
	<=
	1000 * context.value
	&
	mappings.roomDayPatientLoad->filter( //
		element.nodes.d == context.nodes.d //
		& element.nodes.r == context.nodes.r //
		& element.nodes.p.gender == context.nodes.g.name //
	)->sum(element.value)
	+
	patterns.occupantRoomDay->filter( //
		element.nodes.d == context.nodes.d //
		& element.nodes.r == context.nodes.r //
		& element.nodes.o.gender == context.nodes.g.name //
	)->sum(1)
	>=
	context.value
}
```

This is because GIPS calls the respective method in parallel, creating a race condition. Example method (should be generated by the eMoflon::IBeX builder):
```
	/**
	* Creates a new instance of the pattern <code>occupantRoomDay()</code> which does the following:
	* If this pattern is not self-explaining, you really should add some comment in the specification.
	*
	* @return the new instance of the pattern»
	*/
	public OccupantRoomDayPattern occupantRoomDay() {
		try{
			OccupantRoomDayPattern pattern = (OccupantRoomDayPattern) interpreter.getRegisteredGraphTransformationPattern("occupantRoomDay");
			return pattern;
		} catch(Exception e) {
			return new OccupantRoomDayPattern(this, interpreter);
		}
	}
```

Possible workaround/solution: The getters of the eMoflon::IBeX API should be thread-safe (`synchronized`) or GIPS should not call them in parallel.

--- a/org.emoflon.ibex.gt/xtend-src/org/emoflon/ibex/gt/codegen/JavaFileGenerator.xtend
+++ b/org.emoflon.ibex.gt/xtend-src/org/emoflon/ibex/gt/codegen/JavaFileGenerator.xtend
@@ -201,7 +201,7 @@ class JavaFileGenerator {
 						*
 						* @return the new instance of the rule»
 						*/
-						public «getRuleClassName(rule)» «rule.name»(«FOR parameter : rule.parameters SEPARATOR ', '»final «getJavaType(parameter.type)» «parameter.name»Value«ENDFOR») {
+						public synchronized «getRuleClassName(rule)» «rule.name»(«FOR parameter : rule.parameters SEPARATOR ', '»final «getJavaType(parameter.type)» «parameter.name»Value«ENDFOR») {
 							try{
 								«getRuleClassName(rule)» rule = («getRuleClassName(rule)») interpreter.getRegisteredGraphTransformationPattern("«rule.name»");
 								«FOR parameter : rule.parameters»
@@ -212,7 +212,7 @@ class JavaFileGenerator {
 								return new «getRuleClassName(rule)»(this, interpreter«FOR parameter : rule.parameters BEFORE ', 'SEPARATOR ', '»«parameter.name»Value«ENDFOR»);
 							}
 						}
-			«ENDFOR»
+				«ENDFOR»
 				«FOR pattern : ibexModel.patternSet.contextPatterns
 						.filter [ pattern | !rulePreconditions.contains(pattern)]
 						.filter [ pattern | !pattern.name.contains("CONDITION")]»
@@ -222,7 +222,7 @@ class JavaFileGenerator {
 						*
 						* @return the new instance of the pattern»
 						*/
-						public «getPatternClassName(pattern)» «pattern.name»(«FOR parameter : getPatternParameter(pattern) SEPARATOR ', '»final «getJavaType(parameter.type)» «parameter.name»Value«ENDFOR») {
+						public synchronized «getPatternClassName(pattern)» «pattern.name»(«FOR parameter : getPatternParameter(pattern) SEPARATOR ', '»final «getJavaType(parameter.type)» «parameter.name»Value«ENDFOR») {
 							try{
 								«getPatternClassName(pattern)» pattern = («getPatternClassName(pattern)») interpreter.getRegisteredGraphTransformationPattern("«pattern.name»");
 								«FOR parameter : getPatternParameter(pattern)»

